From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik_Tra <erik.traxler@gmx.at>
Date: Sun, 21 Aug 2022 14:44:40 +0200
Subject: [PATCH] 0935-TickFreeze-TickRate


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index fcaaab8ecedce49fca1452302d68a48351ddc42b..a28a0485c837073617dcf2fc3d8625e066b83ce6 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -618,9 +618,12 @@ public class ServerLevel extends Level implements WorldGenLevel {
         }
 
         this.updateSkyBrightness();
-        this.tickTime();
-        gameprofilerfiller.popPush("tickPending");
-        timings.scheduledBlocks.startTiming(); // Paper
+        if (!MinecraftServer.freezed)
+        {
+            this.tickTime();
+            gameprofilerfiller.popPush("tickPending");
+            timings.scheduledBlocks.startTiming(); // Paper
+        }
         if (!this.isDebug()) {
             j = this.getGameTime();
             if (!MinecraftServer.freezed)
@@ -634,21 +637,24 @@ public class ServerLevel extends Level implements WorldGenLevel {
         }
         timings.scheduledBlocks.stopTiming(); // Paper
 
-        gameprofilerfiller.popPush("raid");
-        this.timings.raids.startTiming(); // Paper - timings
-        this.raids.tick();
-        this.timings.raids.stopTiming(); // Paper - timings
-        gameprofilerfiller.popPush("chunkSource");
-        this.timings.chunkProviderTick.startTiming(); // Paper - timings
-        this.getChunkSource().tick(shouldKeepTicking, true);
-        this.timings.chunkProviderTick.stopTiming(); // Paper - timings
-        gameprofilerfiller.popPush("blockEvents");
-        timings.doSounds.startTiming(); // Spigot
-        if (!MinecraftServer.freezed)
-        this.runBlockEvents();
-        timings.doSounds.stopTiming(); // Spigot
-        this.handlingTick = false;
-        gameprofilerfiller.pop();
+                if (!MinecraftServer.freezed)
+                {
+                    gameprofilerfiller.popPush("raid");
+                    this.timings.raids.startTiming(); // Paper - timings
+                    this.raids.tick();
+                    this.timings.raids.stopTiming(); // Paper - timings
+                    gameprofilerfiller.popPush("chunkSource");
+                    this.timings.chunkProviderTick.startTiming(); // Paper - timings
+                    this.getChunkSource().tick(shouldKeepTicking, true);
+                    this.timings.chunkProviderTick.stopTiming(); // Paper - timings
+                    gameprofilerfiller.popPush("blockEvents");
+                    timings.doSounds.startTiming(); // Spigot
+
+                    this.runBlockEvents();
+                    timings.doSounds.stopTiming(); // Spigot
+                    this.handlingTick = false;
+                    gameprofilerfiller.pop();
+                }
         boolean flag = true || !this.players.isEmpty() || !this.getForcedChunks().isEmpty(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players
 
         if (flag) {
@@ -703,10 +709,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
         {
             gameprofilerfiller.push("entityManagement");
             this.entityManager.tick();
+
+            gameprofilerfiller.popPush("gameEvents");
+            this.sendGameEvents();
+            gameprofilerfiller.pop();
         }
-        gameprofilerfiller.popPush("gameEvents");
-        this.sendGameEvents();
-        gameprofilerfiller.pop();
     }
 
     @Override
